{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c36f8e59-525c-4045-9b97-c1e5148932ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "pitman_moore ='''MVPQALLFVPLLGFSLCFGKFPIYTIPDELGPWSPIDIHHLSCPNNLVVEDEGCTNLSEFSYMELKVGYI\n",
    "SAIKVNGFTCTGVVTEAETYTNFVGYVTTTFKRKHFRPTPDACRAAYNWKMAGDPRYEESLHNPYPDYHW\n",
    "LRTVRTTKESLIIISPSVTDLDPYDKSLHSRVFPGGKCSGITVSSTYCSTNHDYTIWMPENPRPRTPCDI\n",
    "FTNSRGKRASKGNKTCGFVDERGLYKSLKGACRLKLCGVLGLRLMDGTWVAMQTSDETKWCPPDQLVNLH\n",
    "DFRSDEIEHLVVEELVKKREECLDALESIMTTKSVSFRRLSHLRKLVPGFGKAYTIFNKTLMEADAHYKS\n",
    "VRTWNEIIPSKGCLKVGGRCHPHVNGVFFNGIILGPDGHVLIPEMQSSLLQQHMELLKSSVIPLMHPLAD\n",
    "PSTVFKEGDEAEDFVEVHLPDVYKQISGVDLGLPNWGKYVLMTAGAMIGLVLIFSLMTWCRRANRPESKQ\n",
    "RSFGGTGRNVSVTSQSGKVIPSWESYKSGGQTRL'''.replace('\\n','')\n",
    "#13\n",
    "vnukovo32G = '''MVPQALLFVPLLVFPLCFGKFPIYTIPDKLGPWSPIDIHHLSCPNNLVVEDEGCTNLSGFSYMELKVGYI\n",
    "LAIKMNGFTCTGVVTEAENYTNFVGYVTTTFKRKHLRPTPDACRAAYNWKMAGDPRYEESLHNPYPDYSW\n",
    "LRTVKTTKESLVIISPSVADLDPYDRSLHSRVFPSGKCSGVAVSSTYCSTNHDYTIWMPENPRLGKSCDI\n",
    "FTNSRGKRASKGSETCGFVDERGLYKSLKGACKLKLCGVLGLRLMDGTWVAMQTSNETKWCPPDQLVNLH\n",
    "DFRSDEIEHLVVEELVRKREECLDALESIMTTKSVSFRRLSHLRKLVPGFGKAYTIFNKTLMEADAHYKS\n",
    "VRTWNEILPSKGCLRVGGRCHPHVNGVFFNGIILGPDGNVLIPEMQSSLLQQHMELLESSVIPLVHPLAD\n",
    "PSTVFKDGDEAEDFVEVHLPDVHNQVSGVDLGLPNWGKYVLLSAGALTALMLIIFLMTCCRRVNRSEPTQ\n",
    "HNLRGTGREVSVTPQTWKIISSWESHKSGGETRL\n",
    "'''.replace('\\n','')\n",
    "#24\n",
    "flurylepG = '''MVPQVLLFVPLLGFSLCFGKFPIYTIPDKLGPWSPIDIHHLSCPNNLVVEDEGCTNLSEFSYMELKVGYI\n",
    "SAIKVNGFTCTGVVTEAETYTNFVGYVTTTFKRKHFRPTPDACRAAYNWKMAGDPRYEESLHNPYPDYHW\n",
    "LRTVKTTKESLVIISPSVTDLDPYDKSLHSRVFPGGNCSGITVSSTYCSTNHDYTIWMPENLRLGTSCDI\n",
    "FTNSRGKRASKGGKTCGFVDERGLYKSLKGACKLKLCGVLGLRLMDGTWVAMQTSDETKWCPPGQLVNLH\n",
    "DFRSDEIEHLVVEELVKKREECLDALESIMTTKSVSFRRLSHLRKLVPGFGKAYTIFNKTLMEADAHYKS\n",
    "VRTWNEIIPSKGCLRVGGRCHPHVNGVFFNGIILGSDGHVLIPEMQSSLLQQHMELLESSVIPLMHPLAD\n",
    "PSTVFKDGDEVEDFVEVHLPDVHEQVSGVELGLPNWGKYVLMIAGALIALMLIIFLMTCCRRVNRPESTQ\n",
    "SSLGETGRNVSVTSQSGKVISSWESYKSGGETRL'''.replace('\\n','')\n",
    "hepfluryG = '''MVPQVLLFAPLLVFPLCFGKFPIYTIPDKLGPWSPIDLHHLSCPNNLVVEDEGCTNLSGFSYMELKVGYISAIKVNGFTCTGVVTEAETYTNFVGYVTTTFKRKHFRPTPDACRAAYNWKMAGDPRYEESLHNPYPDYHWLRTVKTTKESLVIISPSVTDLDPYDKSLHSRVFPGGNCSGITVSSTYCSTNHDYTIWMPENLRLGTSCDIFTHSRGKRASKGDKTCGFVDERGLYKSLKGACKLKLCGVLGLRLMDGTWVAMQTSDETKWCPPGQLVNLHDFRSDEIEHLVEEELVKKREECLDALESIMTTKSVSFRRLSHLRKLVPGFGKAYTIFNKTLMEADAHYKSVQTWNEIIPSKGCLRVGERCHPHVNGVFFNGIILGSDGHVLIPEMQSSLLQQHMELLESSVIPLMHPLADPSTVFKDGDEVEDFVEVHLPDVHKQVSGVDLGLPKWGKYVLMIAGALIALMLIIFLMTCCRRVNRPESTQSNLGGTGRNVSVPSQSGKVISSWESYKSGGETRL'''.replace('\\n','')\n",
    "pasteurG = '''MVPQALLFVPLLVFPLCFGKFPIYTIPDKLGPWSPIDIHHLSCPNNLVVEDEGCTNLSGFSYMELKVGYISAIKMNGFTCTGVVTEAETYTNFVGYVTTTFKRKHFRPTPDACRAAYNWKMAGDPRYEESLHNPYPDYHWLRTVKTTKESLVIISPSVADLDPYDRSLHSRVFPGGNCSGVAVSSTYCSTNHDYTIWMPENPRLGMSCDIFTNSRGKRASKGSETCGFVDERGLYKSLKGACKLKLCGVLGLRLMDGTWVAMQTSNETKWCPPGQLVNLHDFRSDEIEHLVVEELVKKREECLDALESIMTTKSVSFRRLSHLRKLVPGFGKAYTIFNKTLMEADAHYKSVRTWNEIIPSKGCLRVGGRCHPHVNGVFFNGIILGPDGNVLIPEMQSSLLQQHMELLVSSVIPLMHPLADPSTVFKNGDEAEDFVEVHLPDVHERISGVDLGLPNWGKYVLLSAGALTALMLIIFLMTCWRRVNRSEPTQHNLRGTGREVSVTPQSGKIISSWESYKSGGETGL'''.replace('\\n','')\n",
    "hpv35l1 = '''MSLWRSNEATVYLPPVSVSKVVSTDEYVTRTNIYYHAGSSRLLAVGHPYYAIKKQDSNKIAVPKVSGLQYRVFRVKLPDPNKFGFPDTSFYDPASQRLVWACTGVEVGRGQPLGVGISGHPLLNKLDDTENSNKYVGNSGTDNRECISMDYKQTQLCLIGCRPPIGEHWGKGTPCNANQVKAGECPPLELLNTVLQDGDMVDTGFGAMDFTTLQANKSDVPLDICSSICKYPDYLKMVSEPYGDMLFFYLRREQMFVRHLFNRAGTVGETVPADLYIKGTTGTLPSTSYFPTPSGSMVTSDAQIFNKPYWLQRAQGHNNGICWSNQLFVTVVDTTRSTNMSVCSAVSSSDSTYKNDNFKEYLRHGEEYDLQFIFQLCKITLTADVMTYIHSMNPSILEDWNFGLTPPPSGTLEDTYRYVTSQAVTCQKPSAPKPKDDPLKNYTFWEVDLKEKFSADLDQFPLGRKFLLQAGLKARPNFRLGKRAAPASTSKKSSTKRRKVKS'''.replace('\\n','')\n",
    "hpv16_18 ='''MSLWLPSEATVYLPPVPVSKVVSTDEYVARTNIYYHAGTSRLLAVGHPYFPIKKPNNNKILVPKVSGLQYRVFRIHLPDPNKFGFPDTSFYNPDTQRLVWACVGVEVGRGQPLGVGISGHPLLNKLDDTENASAYAANAGVDNRECISMDYKQTQLCLIGCKPPIGEHWGKGSPCTNVAVNPGDCPPLELINTVIQDGDMVDTGFGAMDFTTLQANKSEVPLDICTSICKYPDYIKMVSEPYGDSLFFYLRREQMFVRHLFNRAGAVGENVPDDLYIKGSGSTANLASSNYFPTPSGSMVTSDAQIFNKPYWLQRAQGHNNGICWGNQLFVTVVDTTRSTNMSLCAAISTSETTYKNTNFKEYLRHGEEYDLQFIFQLCKITLTADVMTYIHSMNSTILEDWNFGLQPPPGGTLEDTYRFVTSQAIACQKHTPPAPKEDPLKKYTFWEVNLKEKFSADLDQFPLGRKFLLQAGLKAKPKFTLGKRKATPTTSSTSTTAKRKKRKLMCLYTRVLILHYHLLPLYGPLYHPRPLPLHSILVYMVHIIICGHYIILFLRNVNVFPIFLQMALWRPSDNTVYLPPPSVARVVNTDDYVTPTSIFYHAGSSRLLTVGNPYFRVPAGGGNKQDIPKVSAYQYRVFRVQLPDPNKFGLPDTSIYNPETQRLVWACAGVEIGRGQPLGVGLSGHPFYNKLDDTESSHAATSNVSEDVRDNVSVDYKQTQLCILGCAPAIGEHWAKGTACKSRPLSQGDCPPLELKNTVLEDGDMVDTGYGAMDFSTLQDTKCEVPLDICQSICKYPDYLQMSADPYGDSMFFCLRREQLFARHFWNRAGTMGDTVPQSLYIKGTGMPASPGSCVYSPSPSGSIVTSDSQLFNKPYWLHKAQGHNNGVCWHNQLFVTVVDTTPSTNLTICASTQSPVPGQYDATKFKQYSRHVEEYDLQFIFQLCTITLTADVMSYIHSMNSSILEDWNFGVPPPPTTSLVDTYRFVQSVAITCQKDAAPAENKDPYDKLKFWNVDLKEKFSLDLDQYPLGRKFLVQAGLRRKPTIGPRKRSAPSATTSSKPAKRVRVRARK'''.replace('\\n','')\n",
    "hpv6_11_16_18 ='''MSLWLPSEATVYLPPVPVSKVVSTDEYVARTNIYYHAGTSRLLAVGHPYFPIKKPNNNKILVPKVSGLQYRVFRIHLPDPNKFGFPDTSFYNPDTQRLVWACVGVEVGRGQPLGVGISGHPLLNKLDDTENASAYAANAGVDNRECISMDYKQTQLCLIGCKPPIGEHWGKGSPCTNVAVNPGDCPPLELINTVIQDGDMVDTGFGAMDFTTLQANKSEVPLDICTSICKYPDYIKMVSEPYGDSLFFYLRREQMFVRHLFNRAGAVGENVPDDLYIKGSGSTANLASSNYFPTPSGSMVTSDAQIFNKPYWLQRAQGHNNGICWGNQLFVTVVDTTRSTNMSLCAAISTSETTYKNTNFKEYLRHGEEYDLQFIFQLCKITLTADVMTYIHSMNSTILEDWNFGLQPPPGGTLEDTYRFVTSQAIACQKHTPPAPKEDPLKKYTFWEVNLKEKFSADLDQFPLGRKFLLQAGLKAKPKFTLGKRKATPTTSSTSTTAKRKKRKLMCLYTRVLILHYHLLPLYGPLYHPRPLPLHSILVYMVHIIICGHYIILFLRNVNVFPIFLQMALWRPSDNTVYLPPPSVARVVNTDDYVTPTSIFYHAGSSRLLTVGNPYFRVPAGGGNKQDIPKVSAYQYRVFRVQLPDPNKFGLPDTSIYNPETQRLVWACAGVEIGRGQPLGVGLSGHPFYNKLDDTESSHAATSNVSEDVRDNVSVDYKQTQLCILGCAPAIGEHWAKGTACKSRPLSQGDCPPLELKNTVLEDGDMVDTGYGAMDFSTLQDTKCEVPLDICQSICKYPDYLQMSADPYGDSMFFCLRREQLFARHFWNRAGTMGDTVPQSLYIKGTGMPASPGSCVYSPSPSGSIVTSDSQLFNKPYWLHKAQGHNNGVCWHNQLFVTVVDTTPSTNLTICASTQSPVPGQYDATKFKQYSRHVEEYDLQFIFQLCTITLTADVMSYIHSMNSSILEDWNFGVPPPPTTSLVDTYRFVQSVAITCQKDAAPAENKDPYDKLKFWNVDLKEKFSLDLDQYPLGRKFLVQAGLRRKPTIGPRKRSAPSATTSSKPAKRVRVRARKMWRPSDSTVYVPPPNPVSKVVATDAYVTRTNIFYHASSSRLLAVGHPYFSIKRANKTVVPKVSGYQYRVF\n",
    "KVVLPDPNKFALPDSSLFDPTTQRLVWACTGLEVGRGQPLGVGVSGHPFLNKYDDVENSGSGGNPGQDNR\n",
    "VNVGMDYKQTQLCMVGCAPPLGEHWGKGKQCTNTPVQAGDCPPLELITSVIQDGDMVDTGFGAMNFADLQ\n",
    "TNKSDVPIDICGTTCKYPDYLQMAADPYGDRLFFFLRKEQMFARHFFNRAGEVGEPVPDTLIIKGSGNRT\n",
    "SVGSSIYVNTPSGSLVSSEAQLFNKPYWLQKAQGHNNGICWGNQLFVTVVDTTRSTNMTLCASVTTSSTY\n",
    "TNSDYKEYMRHVEEYDLQFIFQLCSITLSAEVMAYIHTMNPSVLEDWNFGLSPPPNGTLEDTYRYVQSQA\n",
    "ITCQKPTPEKEKPDPYKNLSFWEVNLKEKFSSELDQYPLGRKFLLQSGYRGRSSIRTGVKRPAVSKASAA\n",
    "PKRKRAKTKRMWRPSDSTVYVPPPNPVSKVVATDAYVKRTNIFYHASSSRLLAVGHPYYSIKKVNKTVVPKVSGYQYRVF\n",
    "KVVLPDPNKFALPDSSLFDPTTQRLVWACTGLEVGRGQPLGVGVSGHPLLNKYDDVENSGGYGGNPGQDN\n",
    "RVNVGMDYKQTQLCMVGCAPPLGEHWGKGTQCSNTSVQNGDCPPLELITSVIQDGDMVDTGFGAMNFADL\n",
    "QTNKSDVPLDICGTVCKYPDYLQMAADPYGDRLFFYLRKEQMFARHFFNRAGTVGEPVPDDLLVKGGNNR\n",
    "SSVASSIYVHTPSGSLVSSEAQLFNKPYWLQKAQGHNNGICWGNHLFVTVVDTTRSTNMTLCASVSTSAT\n",
    "YTNSDYKEYMRHVEEFDLQFIFQLCSITLSAEVMAYIHTMNPSVLEDWNFGLSPPPNGTLEDTYRYVQSQ\n",
    "AITCQKPTPEKEKQDPYKDMSFWEVNLKEKFSSELDQFPLGRKFLLQSGYRGRTSARTGIKRPAVSKPST\n",
    "APKRKRTKTKK\n",
    "'''.replace('\\n','')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "03cc4442-254b-4d09-9c56-6ffc408e51ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess\n",
    "from pybatman.functions import train, peptide2index\n",
    "from multiprocessing import Pool\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mpatches\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "adc7be76-05bb-40db-a974-f7fd0802f270",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Population:\n",
    "    \"\"\"A class to represent a population.\n",
    "    Attributes:\n",
    "    path_to_IEPAPI: the absolute path to IEPAPI folder\n",
    "    freq: the absolute path to tsv table with genotype frequencies across population. Tsv file consist of 2 columns without headers. First is a column with absolute counts of each genotype, second is genotypes themselfs\n",
    "    path_to_hla_db: absolute path to table with hla pseudosequences. The format of the table is the same as in IEPAPI table (data/pseudoSequence(ELIM).csv file) \"\"\"\n",
    "    \n",
    "    def __init__(self,path_to_IEPAPI=None,freq=None, path_to_hla_db=None):\n",
    "        if path_to_IEPAPI is None:\n",
    "            path_to_IEPAPI = os.getcwd() + '/IEPAPI/'\n",
    "        if freq is None:\n",
    "            freq = os.getcwd() + '/HLA.tsv'\n",
    "        if path_to_hla_db is None:\n",
    "            self.IEPAPI_db = pd.read_csv(path_to_IEPAPI + 'data/pseudoSequence(ELIM).csv' )\n",
    "        else:\n",
    "            self.IEPAPI_db = pd.read_csv(path_to_hla_db)\n",
    "        if \"tsv\" in freq:\n",
    "            self.genotype_freq = self.pars_tsv(freq)\n",
    "        else:\n",
    "            self.genotype_freq = freq\n",
    "        new_genotype_df = self.genotype_freq\n",
    "        c = 0\n",
    "        for i in range(len(self.genotype_freq)):\n",
    "            genotype = self.genotype_freq['Genotype'][i]\n",
    "            if any(hla not in self.IEPAPI_db['HLA'].tolist() for hla in [('HLA-'+hla[0:7]).replace('*','').replace(':','') for hla in genotype]):\n",
    "                c += self.genotype_freq['Count'][i]\n",
    "                new_genotype_df = new_genotype_df.drop(i)\n",
    "                print(f\"Warning! The genotype {genotype} was removed because it contains unknown alleles\")\n",
    "        new_genotype_df['Freq'] = np.array(new_genotype_df['Freq'])*sum(self.genotype_freq['Count'])/(sum(self.genotype_freq['Count'])-c)\n",
    "        self.genotype_freq = new_genotype_df.reset_index(drop=True)\n",
    "    \n",
    "    \n",
    "    def pars_tsv(self,path_to_tsv):\n",
    "        df = pd.read_csv(path_to_tsv,sep='\\t',names=['ID','HLA'])\n",
    "        ids = np.unique(df['ID'])\n",
    "        freq_d = {'Genotype':[],'Count':[],'Freq':None}\n",
    "        for i in ids:\n",
    "            genotype = df.loc[df['ID'] == i]['HLA'].tolist()\n",
    "            genotype.sort()\n",
    "            if genotype in freq_d['Genotype']:\n",
    "                freq_d['Count'][freq_d['Genotype'].index(genotype)] += 1\n",
    "            else:\n",
    "                freq_d['Genotype'].append(genotype)\n",
    "                freq_d['Count'].append(1)\n",
    "        freq_d['Freq'] = np.array(freq_d['Count'])/len(ids)\n",
    "        return pd.DataFrame(freq_d)\n",
    "    \n",
    "    \n",
    "    def get_individual_sample(self,immun_predictor,N=1,protein=None,mean_number=10):\n",
    "        sample_list = []\n",
    "        for n in range(N):\n",
    "            genotype = np.random.choice(a=np.array(self.genotype_freq['Genotype']), p = self.genotype_freq['Freq'])\n",
    "            sample_list.append(Individual(genotype,immun_predictor,mean_number,protein))\n",
    "        return sample_list\n",
    "            \n",
    "        \n",
    "            \n",
    "            \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "f600c002-9875-4bbf-8778-547ee75c0fd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Protein:\n",
    "    \"\"\"A class to represent a protein\n",
    "    Attributes:\n",
    "    seq: amino acid sequence of protein\"\"\"\n",
    "    \n",
    "    def __init__(self,seq):\n",
    "        self.protein_seq = seq\n",
    "        self.peptide_pool = self.get_peptide_pool()\n",
    "    \n",
    "    def get_peptide_pool(self,seq=None,k=9, peptide_weight=None):\n",
    "        peptide_list = []\n",
    "        if seq is None:\n",
    "            seq = self.protein_seq\n",
    "        for i in range(len(seq)-k):\n",
    "            peptide_list.append(seq[i:i+k])\n",
    "        if peptide_weight is not None:\n",
    "            peptide_list = np.random.choice(a = peptide_list,p = peptide_weight)\n",
    "        return peptide_list\n",
    "\n",
    "# разные режимы для генерации пептидов\n",
    "# у каждого пептида свой вес"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "ebfbc1d5-be63-48d0-9fa3-d72ebb48e82d",
   "metadata": {},
   "outputs": [],
   "source": [
    "class peptide_MHC_complex:\n",
    "    \"\"\"A class to represent pMHC complex\n",
    "    Attributes:\n",
    "    peptide: peptide sequence\n",
    "    HLA: HLA name\n",
    "    pres: presentation score for peptide-HLA pair\n",
    "    immun: immunogenecity score for peptide-HLA pair\n",
    "    HLA_i: HLA index\"\"\"\n",
    "    \n",
    "    def __init__(self,peptide,HLA,pres,immun,\n",
    "                 HLA_i):\n",
    "        self.peptide = peptide\n",
    "        self.HLA = HLA\n",
    "        self.pres = pres\n",
    "        self.immun = immun\n",
    "        self.HLA_i = HLA_i\n",
    "    \n",
    "    def cross (self,cross_react_predictor,mutant_pMHC):\n",
    "        homology = cross_react_predictor.predict([self.peptide],[mutant_pMHC.peptide])\n",
    "        c=0\n",
    "        for i in range(len(self.peptide)):\n",
    "            if self.peptide[i] != mutant_pMHC.peptide[i]:\n",
    "                c += 1\n",
    "        if c > 2:\n",
    "            homology = 0\n",
    "        homology = homology*mutant_pMHC.pres/self.pres\n",
    "        if homology > 1:\n",
    "            homology = 1\n",
    "        return homology"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "5e312bcb-256c-4ce2-8e55-e4416d5364c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Individual:\n",
    "    \"\"\"A class to represent a individual\n",
    "    Attributes:\n",
    "    genotype: list with HLA alleles\n",
    "    immun_predictor: object of the class IEPAPI_immunogenicity_predictor\n",
    "    mean_number: mean number of pMHC complexes that will be presented\n",
    "    protein_wild: object of the class Protein that represented a wild protein\n",
    "    \"\"\"\n",
    "    \n",
    "    def __init__(self,genotype,immun_predictor,mean_number, protein_wild=None):\n",
    "        self.genotype = genotype\n",
    "        self.protein_wild = protein_wild\n",
    "        if self.protein_wild is not None:\n",
    "            self.pMHC_pool = self.pMHC_selection(self.get_pMHC_pool(self.protein_wild,immun_predictor),mean_number)\n",
    "        \n",
    "    def get_pMHC_pool(self,protein, imumn_predictor):\n",
    "        pep_list = protein.peptide_pool\n",
    "        pMHC_list = []\n",
    "        pMHC_d = {'HLA':[],'peptide':[],'pres':[],'immun':[],'HLA_i':[]}\n",
    "        for j,hla in enumerate(self.genotype):\n",
    "            pMHC_d['HLA_i'] += [j]*len(pep_list)\n",
    "            pMHC_d['HLA'] += [hla]*len(pep_list)\n",
    "            pMHC_d['peptide'] += pep_list\n",
    "        res = imumn_predictor.predict(pep_list,self.genotype)\n",
    "        pMHC_d['pres'] = res[1]\n",
    "        pMHC_d['immun'] = res[0]\n",
    "        for i in range(len(pMHC_d['peptide'])):\n",
    "            pMHC_list.append(peptide_MHC_complex(pMHC_d['peptide'][i],pMHC_d['HLA'][i],pMHC_d['pres'][i],pMHC_d['immun'][i],pMHC_d['HLA_i'][i]))\n",
    "        return pMHC_list\n",
    "                \n",
    "    def set_protein_wild(self, protein_wild,immun_predictor,mean_number=20):\n",
    "        self.protein_wild = protein_wild\n",
    "        self.pMHC_pool = self.pMHC_selection(self.get_pMHC_pool(self.protein_wild,immun_predictor),mean_number)\n",
    "\n",
    "        \n",
    "    def pMHC_selection(self,pMHC,mean_number):\n",
    "        P=[]\n",
    "        for pMHC_i in pMHC:\n",
    "            P.append(pMHC_i.pres*pMHC_i.immun)\n",
    "        P = np.array(P)\n",
    "        if mean_number is not None:\n",
    "            P = P-((sum(P)-mean_number)/len(P))\n",
    "            P[P<0] = 0\n",
    "            P[P>1] = 1\n",
    "        pMHC_selected = []\n",
    "        for i in range(len(P)):\n",
    "            if np.random.choice(a=[True,False], p=[P[i],1-P[i]]):\n",
    "                pMHC_selected.append(pMHC[i])\n",
    "        return pMHC_selected\n",
    "        \n",
    "\n",
    "    def cross_react(self,cross_react_predictor,immun_predictor,mutant_protein):\n",
    "        result = []\n",
    "        wild_pMHC = self.pMHC_pool\n",
    "        if len(wild_pMHC) == 0:\n",
    "            return 0\n",
    "        mutant_pMHC = self.get_pMHC_pool(mutant_protein,immun_predictor)\n",
    "        for w_pMHC in wild_pMHC:\n",
    "            for m_pMHC in mutant_pMHC:\n",
    "                if w_pMHC.HLA_i == m_pMHC.HLA_i:\n",
    "                    result = np.append(result, w_pMHC.cross(cross_react_predictor,m_pMHC))\n",
    "        return sum(result)/len(wild_pMHC)\n",
    "        \n",
    "        \n",
    "        \n",
    "        \n",
    "        \n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "76985f85-987b-4c37-a1a8-af8ad82df2d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "class BATMAN_predictor:\n",
    "    \"\"\"A class to BATMAN tool\n",
    "    Attributes:\n",
    "    path_to_BATMAN: the absolute path to BATMAN folder (pybatman package)\"\"\"\n",
    "    inferred_AA_matrix, weight_profile, path_to_BATMAN = None,None,None\n",
    "    max_value = 4.382771069923114\n",
    "    \n",
    "    def __init__(self, path_to_BATMAN=None, path_to_train_data=None):\n",
    "        \n",
    "        if path_to_BATMAN is None:\n",
    "            result = subprocess.run(\n",
    "            ['conda', 'list', '-n', 'BATMAN'],\n",
    "            capture_output=True,\n",
    "            text=True,\n",
    "            check=True)\n",
    "            for line in result.stdout.splitlines():\n",
    "                if 'pybatman' in line:\n",
    "                    parts = line.split()\n",
    "                    package_path = os.path.join(result.stdout.splitlines()[0].split()[5][:-1], 'lib/python3.11/site-packages/',parts[0])\n",
    "                    self.path_to_BATMAN = package_path\n",
    "        else:\n",
    "            self.path_to_BATMAN = path_to_BATMAN\n",
    "\n",
    "        if path_to_train_data is None:\n",
    "            path_to_train_data = os.getcwd() + '/test_input_BATMAN'\n",
    "            \n",
    "        full_peptide_data = pd.read_csv(path_to_train_data)\n",
    "        full_peptide_data.tcr = 'TCR1'\n",
    "        full_peptide_data.to_csv('BATMAN_train.csv')\n",
    "#AA_matrix_name = 'blosum100' \n",
    "        AA_matrix_prior = pd.read_csv(self.path_to_BATMAN+ '/data/AA_matrices/blosum100.csv',index_col=0)\n",
    "        inferred_weights, inferred_AA_matrix = train('BATMAN_train.csv','full', AA_matrix_prior, steps=80000, seed=10)\n",
    "        weight_profile = inferred_weights.to_numpy()\n",
    "        self.weight_profile = weight_profile\n",
    "        self.inferred_AA_matrix = inferred_AA_matrix\n",
    "    \n",
    "    def predict(self,wild_peptides,mutant_peptides):\n",
    "        peptide_distance = peptide2index(wild_peptides,\n",
    "                                mutant_peptides,\n",
    "                                 self.inferred_AA_matrix,\n",
    "                                 self.weight_profile)\n",
    "        peptide_distance = np.array(peptide_distance)\n",
    "        return self.norm_distance(peptide_distance)\n",
    "\n",
    "\n",
    "    def norm_distance(self,peptide_distance):\n",
    "        peptide_distance[peptide_distance > self.max_value] = self.max_value\n",
    "        peptide_distance[peptide_distance < 0] = 0\n",
    "        peptide_distance = peptide_distance/self.max_value\n",
    "        peptide_distance = np.ones(len(peptide_distance)) - peptide_distance\n",
    "        return peptide_distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "04d7aeeb-4fb1-42f3-b7a3-5f8074388080",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "class IEPAPI_immunogenicity_predictor:\n",
    "    \"\"\"A class to IEPAPI tool\n",
    "    Attributes:\n",
    "    path_to_IEPAPI: the absolute path to IEPAPI folder\n",
    "    python_path: the absolute path to python in IEPAPI environment\n",
    "    hla_pseudoseq_data: the absolute path to HLA pseudosequence table\"\"\"\n",
    "    hla_pseudoseq_data = None\n",
    "    log = {}\n",
    "    path_to_IEPAPI  = None\n",
    "    python_path = None\n",
    "    def __init__(self, path_to_IEPAPI=None, path_to_hla_db=None, python_path=None):\n",
    "        if path_to_IEPAPI is None:\n",
    "            self.path_to_IEPAPI = os.getcwd() + '/IEPAPI/'\n",
    "        else:\n",
    "            self.path_to_IEPAPI = path_to_IEPAPI\n",
    "        if python_path is None:\n",
    "            result = subprocess.run(\n",
    "    ['conda', 'list', '-n', 'IEPAPI'],\n",
    "            capture_output=True,\n",
    "            text=True,\n",
    "            check=True)\n",
    "            self.python_path = result.stdout.splitlines()[0].split()[5][:-1]+'/bin/python3.7'\n",
    "        else:\n",
    "            self.python_path = python_path\n",
    "        if path_to_hla_db is None:\n",
    "            self.IEPAPI_db = self.path_to_IEPAPI+'data/pseudoSequence(ELIM).csv'\n",
    "        else:\n",
    "            self.IEPAPI_db = path_to_hla_db\n",
    "        self.hla_pseudoseq_data = pd.read_csv(self.IEPAPI_db)\n",
    "        \n",
    "        \n",
    "    def predict(self,peptides,hla_vector):\n",
    "        immun = np.array([-1.0]*(len(peptides)*len(hla_vector)))\n",
    "        pres = np.array([-1.0]*(len(peptides)*len(hla_vector)))\n",
    "        with open(self.path_to_IEPAPI+'IEPAPI_input', 'w') as file:\n",
    "            file.write('peptide,HLA,seq')\n",
    "            file.write('\\n')\n",
    "            for j,hla in enumerate(hla_vector):\n",
    "                hla = 'HLA-'+ hla\n",
    "                hla = hla[0:11]\n",
    "                hla_to_df = hla.replace('*','').replace(':','')\n",
    "                hla_pseudoseq = self.hla_pseudoseq_data.loc[self.hla_pseudoseq_data['HLA'] == hla_to_df,'pseudoSeq'].iloc[0]\n",
    "                for i in range(len(peptides)):\n",
    "                    if peptides[i]+hla in self.log:\n",
    "                        log_pep = self.log[peptides[i]+hla]\n",
    "                        pres[i+len(peptides)*j] = log_pep[0]\n",
    "                        immun[i+len(peptides)*j] = log_pep[1]\n",
    "                    else:\n",
    "                        file.write(f'{peptides[i]},{hla},{hla_pseudoseq}')\n",
    "                        file.write('\\n')\n",
    "        if all(immun != -1):\n",
    "            return (immun, pres)\n",
    "        subprocess.run([self.python_path, self.path_to_IEPAPI+'IEPAPI_predict.py', '--input', 'IEPAPI_input', '--output','IEPAPI_output'],  cwd=self.path_to_IEPAPI)\n",
    "        out = pd.read_csv(self.path_to_IEPAPI+'IEPAPI_output')\n",
    "        for j,i in enumerate(np.where(immun==-1.0)[0]):\n",
    "            immun[i] = out.iloc[j,3]\n",
    "            pres[i] = out.iloc[j,2]\n",
    "            self.log[out.iloc[j,0]+out.iloc[j,1]] = (out.iloc[j,2],out.iloc[j,3])\n",
    "        return (immun, pres)\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "5f68c139-3f90-49e9-a240-187cca196fc2",
   "metadata": {
    "editable": true,
    "scrolled": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "from mhctools import MHCflurry\n",
    "def baseline(protein1,protein2,a):\n",
    "    for j,i in enumerate(a):\n",
    "        if len(i) > 7:\n",
    "            a[j]=a[j][:7]\n",
    "    predictor = MHCflurry(alleles=a)\n",
    "    protein_sequences = {\n",
    "    \"Protein1\": protein1.protein_seq,\n",
    "    \"Protein2\": protein2.protein_seq}\n",
    "    binding_predictions = predictor.predict_subsequences(protein_sequences, peptide_lengths=[9])\n",
    "    df = binding_predictions.to_dataframe()\n",
    "    df = df.loc[df['percentile_rank']<=2,].reset_index(drop=True)\n",
    "    var1 = df.loc[df['source_sequence_name']==\"Protein1\",\"peptide\"].tolist()\n",
    "    var2 = df.loc[df['source_sequence_name']==\"Protein2\",\"peptide\"].tolist()\n",
    "    c = 0\n",
    "    for v in var2:\n",
    "        if v in var1:\n",
    "            c +=1\n",
    "    return(c/len(var2))\n",
    "    \n",
    "def compare(a,b):\n",
    "    X_samples = np.random.choice(a,1000)\n",
    "    Y_samples = np.random.choice(b, 1000)\n",
    "    return np.mean(X_samples > Y_samples)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "b5b34948-e06e-462d-b9a0-bb0fd1cd78b4",
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "'>' not supported between instances of 'int' and 'NoneType'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mTypeError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[75]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m \u001b[38;5;28;43mmax\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[32;43m12\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "\u001b[31mTypeError\u001b[39m: '>' not supported between instances of 'int' and 'NoneType'"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "eac579f3-2337-4016-819d-bdc94383cbae",
   "metadata": {
    "editable": true,
    "scrolled": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "Pop1 = Population()\n",
    "immun_predictor = IEPAPI_immunogenicity_predictor()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "2bab2813-127e-4d6a-a1a4-bbec0498110c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/orlov.av/anaconda3/envs/BATMAN/lib/python3.11/site-packages/pymc/distributions/transforms.py:53: FutureWarning: univariate_ordered has been deprecated, use ordered instead.\n",
      "  warnings.warn(f\"{name} has been deprecated, use ordered instead.\", FutureWarning)\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "ebcaf2948692477da9b72752c280d9a9",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Output()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"></pre>\n"
      ],
      "text/plain": []
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Shape validation failed: input_shape: (1, 50000), minimum_shape: (chains=2, draws=4)\n"
     ]
    }
   ],
   "source": [
    "cross_react_predictor = BATMAN_predictor()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "7a6acb60-29a8-4522-af42-b2dc98193f2c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 2/2 [00:04<00:00,  2.14s/it]\n",
      "100%|██████████| 1/1 [00:00<00:00,  3.28it/s]\n",
      "100%|██████████| 2/2 [00:04<00:00,  2.00s/it]\n",
      "100%|██████████| 1/1 [00:00<00:00,  2.72it/s]\n",
      "100%|██████████| 2/2 [00:03<00:00,  1.69s/it]\n",
      "100%|██████████| 1/1 [00:00<00:00,  3.20it/s]\n",
      "100%|██████████| 1/1 [00:02<00:00,  2.19s/it]\n",
      "100%|██████████| 1/1 [00:00<00:00,  4.84it/s]\n",
      "100%|██████████| 2/2 [00:04<00:00,  2.09s/it]\n",
      "100%|██████████| 1/1 [00:00<00:00,  3.19it/s]\n",
      "100%|██████████| 2/2 [00:03<00:00,  1.64s/it]\n",
      "100%|██████████| 1/1 [00:00<00:00,  3.11it/s]\n",
      "100%|██████████| 1/1 [00:00<00:00,  1.05it/s]\n",
      "100%|██████████| 1/1 [00:00<00:00,  3.27it/s]\n",
      "Process ForkPoolWorker-309:\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mKeyboardInterrupt\u001b[39m                         Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[152]\u001b[39m\u001b[32m, line 68\u001b[39m\n\u001b[32m     62\u001b[39m     plt.show()\n\u001b[32m     66\u001b[39m \u001b[38;5;66;03m#Cross_react_predict(Protein(protein_A), Protein(protein_B), Pop1, immun_predictor,cross_react_predictor,N_iterations=50,k=10,N_individuals=4)\u001b[39;00m\n\u001b[32m---> \u001b[39m\u001b[32m68\u001b[39m \u001b[43mCross_react_plot\u001b[49m\u001b[43m(\u001b[49m\u001b[43mProtein\u001b[49m\u001b[43m(\u001b[49m\u001b[43mprotein_A\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mProtein\u001b[49m\u001b[43m(\u001b[49m\u001b[43mprotein_B\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mProtein\u001b[49m\u001b[43m(\u001b[49m\u001b[43mprotein_C\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mPop1\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mimmun_predictor\u001b[49m\u001b[43m,\u001b[49m\u001b[43mcross_react_predictor\u001b[49m\u001b[43m,\u001b[49m\u001b[43mN_iterations\u001b[49m\u001b[43m=\u001b[49m\u001b[32;43m40\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43mk\u001b[49m\u001b[43m=\u001b[49m\u001b[32;43m10\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43mN_individuals\u001b[49m\u001b[43m=\u001b[49m\u001b[32;43m4\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[152]\u001b[39m\u001b[32m, line 28\u001b[39m, in \u001b[36mCross_react_plot\u001b[39m\u001b[34m(protein1, protein2, protein3, genotypes, immun_predictor, cross_react_predictor, N_iterations, k, N_individuals, savefile)\u001b[39m\n\u001b[32m     26\u001b[39m c2 = \u001b[33m'\u001b[39m\u001b[33mblue\u001b[39m\u001b[33m'\u001b[39m\n\u001b[32m     27\u001b[39m data1 = Cross_react_predict(protein1, protein2, genotypes, immun_predictor,cross_react_predictor,N_iterations,k,N_individuals)\n\u001b[32m---> \u001b[39m\u001b[32m28\u001b[39m data2 = \u001b[43mCross_react_predict\u001b[49m\u001b[43m(\u001b[49m\u001b[43mprotein1\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mprotein3\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgenotypes\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mimmun_predictor\u001b[49m\u001b[43m,\u001b[49m\u001b[43mcross_react_predictor\u001b[49m\u001b[43m,\u001b[49m\u001b[43mN_iterations\u001b[49m\u001b[43m,\u001b[49m\u001b[43mk\u001b[49m\u001b[43m,\u001b[49m\u001b[43mN_individuals\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m     29\u001b[39m plt.figure()\n\u001b[32m     30\u001b[39m plt.boxplot(data1,positions=x-\u001b[32m100\u001b[39m,widths=\u001b[32m150\u001b[39m,notch=\u001b[38;5;28;01mTrue\u001b[39;00m, patch_artist=\u001b[38;5;28;01mTrue\u001b[39;00m,\n\u001b[32m     31\u001b[39m         boxprops=\u001b[38;5;28mdict\u001b[39m(facecolor=\u001b[33m'\u001b[39m\u001b[33mpink\u001b[39m\u001b[33m'\u001b[39m, color=c1),\n\u001b[32m     32\u001b[39m         capprops=\u001b[38;5;28mdict\u001b[39m(color=c1),\n\u001b[32m     33\u001b[39m         whiskerprops=\u001b[38;5;28mdict\u001b[39m(color=c1),\n\u001b[32m     34\u001b[39m         flierprops=\u001b[38;5;28mdict\u001b[39m(color=c1, markeredgecolor=c1),\n\u001b[32m     35\u001b[39m         medianprops=\u001b[38;5;28mdict\u001b[39m(color=c1))\n",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[152]\u001b[39m\u001b[32m, line 16\u001b[39m, in \u001b[36mCross_react_predict\u001b[39m\u001b[34m(protein1, protein2, genotypes, immun_predictor, cross_react_predictor, N_iterations, k, N_individuals)\u001b[39m\n\u001b[32m     14\u001b[39m         individuals.append((Individual(genotype,immun_predictor,k, protein1)))\n\u001b[32m     15\u001b[39m pool = Pool(processes=\u001b[32m1\u001b[39m)\n\u001b[32m---> \u001b[39m\u001b[32m16\u001b[39m result = \u001b[43mpool\u001b[49m\u001b[43m.\u001b[49m\u001b[43mstarmap\u001b[49m\u001b[43m(\u001b[49m\u001b[43mind_dist\u001b[49m\u001b[43m,\u001b[49m\u001b[43m[\u001b[49m\u001b[43m(\u001b[49m\u001b[43mind\u001b[49m\u001b[43m,\u001b[49m\u001b[43mcross_react_predictor\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mimmun_predictor\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mprotein1\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mprotein2\u001b[49m\u001b[43m,\u001b[49m\u001b[43mk\u001b[49m\u001b[43m,\u001b[49m\u001b[43mN_iterations\u001b[49m\u001b[43m)\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mfor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mind\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mindividuals\u001b[49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m     17\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m result\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/anaconda3/envs/BATMAN/lib/python3.11/multiprocessing/pool.py:375\u001b[39m, in \u001b[36mPool.starmap\u001b[39m\u001b[34m(self, func, iterable, chunksize)\u001b[39m\n\u001b[32m    369\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34mstarmap\u001b[39m(\u001b[38;5;28mself\u001b[39m, func, iterable, chunksize=\u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[32m    370\u001b[39m \u001b[38;5;250m    \u001b[39m\u001b[33;03m'''\u001b[39;00m\n\u001b[32m    371\u001b[39m \u001b[33;03m    Like `map()` method but the elements of the `iterable` are expected to\u001b[39;00m\n\u001b[32m    372\u001b[39m \u001b[33;03m    be iterables as well and will be unpacked as arguments. Hence\u001b[39;00m\n\u001b[32m    373\u001b[39m \u001b[33;03m    `func` and (a, b) becomes func(a, b).\u001b[39;00m\n\u001b[32m    374\u001b[39m \u001b[33;03m    '''\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m375\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_map_async\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfunc\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43miterable\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstarmapstar\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mchunksize\u001b[49m\u001b[43m)\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/anaconda3/envs/BATMAN/lib/python3.11/multiprocessing/pool.py:768\u001b[39m, in \u001b[36mApplyResult.get\u001b[39m\u001b[34m(self, timeout)\u001b[39m\n\u001b[32m    767\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34mget\u001b[39m(\u001b[38;5;28mself\u001b[39m, timeout=\u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[32m--> \u001b[39m\u001b[32m768\u001b[39m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mwait\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    769\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mself\u001b[39m.ready():\n\u001b[32m    770\u001b[39m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTimeoutError\u001b[39;00m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/anaconda3/envs/BATMAN/lib/python3.11/multiprocessing/pool.py:765\u001b[39m, in \u001b[36mApplyResult.wait\u001b[39m\u001b[34m(self, timeout)\u001b[39m\n\u001b[32m    764\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34mwait\u001b[39m(\u001b[38;5;28mself\u001b[39m, timeout=\u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[32m--> \u001b[39m\u001b[32m765\u001b[39m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_event\u001b[49m\u001b[43m.\u001b[49m\u001b[43mwait\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/anaconda3/envs/BATMAN/lib/python3.11/threading.py:629\u001b[39m, in \u001b[36mEvent.wait\u001b[39m\u001b[34m(self, timeout)\u001b[39m\n\u001b[32m    627\u001b[39m signaled = \u001b[38;5;28mself\u001b[39m._flag\n\u001b[32m    628\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m signaled:\n\u001b[32m--> \u001b[39m\u001b[32m629\u001b[39m     signaled = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_cond\u001b[49m\u001b[43m.\u001b[49m\u001b[43mwait\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    630\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m signaled\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/anaconda3/envs/BATMAN/lib/python3.11/threading.py:327\u001b[39m, in \u001b[36mCondition.wait\u001b[39m\u001b[34m(self, timeout)\u001b[39m\n\u001b[32m    325\u001b[39m \u001b[38;5;28;01mtry\u001b[39;00m:    \u001b[38;5;66;03m# restore state no matter what (e.g., KeyboardInterrupt)\u001b[39;00m\n\u001b[32m    326\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m timeout \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[32m--> \u001b[39m\u001b[32m327\u001b[39m         \u001b[43mwaiter\u001b[49m\u001b[43m.\u001b[49m\u001b[43macquire\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    328\u001b[39m         gotit = \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[32m    329\u001b[39m     \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "\u001b[31mKeyboardInterrupt\u001b[39m: "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/home/orlov.av/anaconda3/envs/BATMAN/lib/python3.11/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/orlov.av/anaconda3/envs/BATMAN/lib/python3.11/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/orlov.av/anaconda3/envs/BATMAN/lib/python3.11/multiprocessing/pool.py\", line 125, in worker\n",
      "    result = (True, func(*args, **kwds))\n",
      "                    ^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/orlov.av/anaconda3/envs/BATMAN/lib/python3.11/multiprocessing/pool.py\", line 51, in starmapstar\n",
      "    return list(itertools.starmap(args[0], args[1]))\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/tmp/ipykernel_3000975/1477629000.py\", line 4, in ind_dist\n",
      "    individual.set_protein_wild(protein1, immun_predictor)\n",
      "  File \"/tmp/ipykernel_3000975/684882589.py\", line 33, in set_protein_wild\n",
      "    self.pMHC_pool = self.pMHC_selection(self.get_pMHC_pool(self.protein_wild,immun_predictor),mean_number)\n",
      "                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/tmp/ipykernel_3000975/684882589.py\", line 24, in get_pMHC_pool\n",
      "    res = imumn_predictor.predict(pep_list,self.genotype)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/tmp/ipykernel_3000975/2054101993.py\", line 37, in predict\n",
      "    hla_pseudoseq = self.hla_pseudoseq_data.loc[self.hla_pseudoseq_data['HLA'] == hla_to_df,'pseudoSeq'].iloc[0]\n",
      "                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/orlov.av/anaconda3/envs/BATMAN/lib/python3.11/site-packages/pandas/core/indexing.py\", line 1184, in __getitem__\n",
      "    return self._getitem_tuple(key)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/orlov.av/anaconda3/envs/BATMAN/lib/python3.11/site-packages/pandas/core/indexing.py\", line 1368, in _getitem_tuple\n",
      "    return self._getitem_lowerdim(tup)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/orlov.av/anaconda3/envs/BATMAN/lib/python3.11/site-packages/pandas/core/indexing.py\", line 1089, in _getitem_lowerdim\n",
      "    return getattr(section, self.name)[new_key]\n",
      "           ~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^\n",
      "  File \"/home/orlov.av/anaconda3/envs/BATMAN/lib/python3.11/site-packages/pandas/core/indexing.py\", line 1191, in __getitem__\n",
      "    return self._getitem_axis(maybe_callable, axis=axis)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/orlov.av/anaconda3/envs/BATMAN/lib/python3.11/site-packages/pandas/core/indexing.py\", line 1413, in _getitem_axis\n",
      "    return self._getbool_axis(key, axis=axis)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/orlov.av/anaconda3/envs/BATMAN/lib/python3.11/site-packages/pandas/core/indexing.py\", line 1209, in _getbool_axis\n",
      "    key = check_bool_indexer(labels, key)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/orlov.av/anaconda3/envs/BATMAN/lib/python3.11/site-packages/pandas/core/indexing.py\", line 2632, in check_bool_indexer\n",
      "    def check_bool_indexer(index: Index, key) -> np.ndarray:\n",
      "    \n",
      "KeyboardInterrupt\n",
      "100%|██████████| 2/2 [00:04<00:00,  2.20s/it]\n",
      "100%|██████████| 1/1 [00:00<00:00,  2.55it/s]\n"
     ]
    }
   ],
   "source": [
    "def ind_dist(individual, cross_react_predictor,immun_predictor,protein1,protein2,k, N_iterations):\n",
    "    result = []\n",
    "    for i in range(N_iterations):\n",
    "        individual.set_protein_wild(protein1, immun_predictor)\n",
    "        result.append(individual.cross_react(cross_react_predictor, immun_predictor,protein2))\n",
    "    return result\n",
    "    \n",
    "def Cross_react_predict(protein1, protein2, genotypes, immun_predictor,cross_react_predictor,N_iterations=50,k=10,N_individuals=None):\n",
    "    if type(genotypes).__name__ == 'Population':\n",
    "        individuals = genotypes.get_individual_sample(immun_predictor,N=N_individuals,mean_number=k)\n",
    "    else:\n",
    "        individuals = []\n",
    "        for genotype in genotypes:\n",
    "            individuals.append((Individual(genotype,immun_predictor,k, protein1)))\n",
    "    pool = Pool(processes=1)\n",
    "    result = pool.starmap(ind_dist,[(ind,cross_react_predictor, immun_predictor, protein1, protein2,k,N_iterations) for ind in individuals])\n",
    "    return result\n",
    "def Cross_react_plot(protein1, protein2, protein3, genotypes, immun_predictor,cross_react_predictor,N_iterations=50,k=10,N_individuals=None, savefile=None):\n",
    "    if N_individuals is None:\n",
    "        x_len = len(genotypes)\n",
    "    else:\n",
    "        x_len = N_individuals\n",
    "        \n",
    "    x = np.array([i for i in range(500, 500*x_len+500,500)])\n",
    "    c1 = \"red\"\n",
    "    c2 = 'blue'\n",
    "    data1 = Cross_react_predict(protein1, protein2, genotypes, immun_predictor,cross_react_predictor,N_iterations,k,N_individuals)\n",
    "    data2 = Cross_react_predict(protein1, protein3, genotypes, immun_predictor,cross_react_predictor,N_iterations,k,N_individuals)\n",
    "    plt.figure()\n",
    "    plt.boxplot(data1,positions=x-100,widths=150,notch=True, patch_artist=True,\n",
    "            boxprops=dict(facecolor='pink', color=c1),\n",
    "            capprops=dict(color=c1),\n",
    "            whiskerprops=dict(color=c1),\n",
    "            flierprops=dict(color=c1, markeredgecolor=c1),\n",
    "            medianprops=dict(color=c1))\n",
    "    plt.boxplot(data2,positions=x+100,widths=150,notch=True, patch_artist=True,\n",
    "            boxprops=dict(facecolor='cyan', color=c2),\n",
    "            capprops=dict(color=c2),\n",
    "            whiskerprops=dict(color=c2),\n",
    "            flierprops=dict(color=c2, markeredgecolor=c2),\n",
    "            medianprops=dict(color=c2))\n",
    "#for j,i in enumerate(x):\n",
    "#    plt.hlines(y = result_baseline_mp_ac[j], xmin=i+25,xmax=i+175, color = 'black')\n",
    "#    plt.hlines(y = result_baseline_mp_ab[j], xmin=i-175,xmax=i-25, color = 'black')\n",
    "    plt.xlim(100,500*x_len+500)\n",
    "    plt.title('Cross-reactivity boxplots')\n",
    "    plt.xticks(x,[str(i) for i in range(x_len)])\n",
    "    plt.xlabel('Genotype')\n",
    "    plt.ylabel('Cross-reactivity')\n",
    "    blue_patch = mpatches.Patch(color='blue', label='C protein')\n",
    "    red_patch = mpatches.Patch(color='red', label='B protein')  \n",
    "    plt.legend(handles=[blue_patch, red_patch])\n",
    "#plt.hlines(y = 0.9867, xmin=100,xmax=6000, color = 'red', linestyles='--')\n",
    "#plt.hlines(y = 0.9867, xmin=100,xmax=6000, color = 'blue', linestyles='--')\n",
    "#plt.hlines(y = result_base_ab_v, xmin=250,xmax=6000, color = 'red', linestyles='--')\n",
    "#plt.hlines(y = result_base_ac_v, xmin=250,xmax=6000, color = 'blue', linestyles='--')\n",
    "#plt.xticks([1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,],['5','10','15','20','25','30','40','50','60','70','80','90','100'])\n",
    "    #plt.savefig('valid_pair.png')\n",
    "    if savefile is not None:\n",
    "        plt.savefig(savefile)\n",
    "        return\n",
    "    plt.show()\n",
    "\n",
    "    \n",
    "    \n",
    "#Cross_react_predict(Protein(protein_A), Protein(protein_B), Pop1, immun_predictor,cross_react_predictor,N_iterations=50,k=10,N_individuals=4)\n",
    "        \n",
    "Cross_react_plot(Protein(protein_A), Protein(protein_B), Protein(protein_C), Pop1, immun_predictor,cross_react_predictor,N_iterations=400,k=10,N_individuals=4)\n",
    "\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "5e94a342-abfb-4a50-bad5-0f9a51958450",
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_A ='''MVPQALLFVPLLGFSLCFGKFPIYTIPDELGPWSPIDIHHLSCPNNLVVEDEGCTNLSEFSYMELKVGYI\n",
    "SAIKVNGFTCTGVVTEAETYTNFVGYVTTTFKRKHFRPTPDACRAAYNWKMAGDPRYEESLHNPYPDYHW\n",
    "LRTVRTTKESLIIISPSVTDLDPYDKSLHSRVFPGGKCSGITVSSTYCSTNHDYTIWMPENPRPRTPCDI\n",
    "FTNSRGKRASKGNKTCGFVDERGLYKSLKGACRLKLCGVLGLRLMDGTWVAMQTSDETKWCPPDQLVNLH\n",
    "DFRSDEIEHLVVEELVKKREECLDALESIMTTKSVSFRRLSHLRKLVPGFGKAYTIFNKTLMEADAHYKS\n",
    "VRTWNEIIPSKGCLKVGGRCHPHVNGVFFNGIILGPDGHVLIPEMQSSLLQQHMELLKSSVIPLMHPLAD\n",
    "PSTVFKEGDEAEDFVEVHLPDVYKQISGVDLGLPNWGKYVLMTAGAMIGLVLIFSLMTWCRRANRPESKQ\n",
    "RSFGGTGRNVSVTSQSGKVIPSWESYKSGGQTRL'''.replace('\\n','')\n",
    "protein_B ='''EVPQALLFVPLLGFSLCFGKFPIYTIPDELGPWSPIDIHHLSCPNNLVVEDEGCTNLSEFSYMELKVGYI\n",
    "SAIKVNGFTCTGVVTEAETYTNFVGYVTTTFKRKHFRPTPDDCRAAYNWKMAGDPRYEESLHNPYPDYHW\n",
    "LRTVRTTKESLIIISPSVTDLDPYDKSLHSRVFPGGKCSGITVSSTYCSTNHDYTIWMPENPPRTPCDI\n",
    "FTNSRGKRASKGNKTCGFVDERGVYKSLKGACRLKLCGVLGLRLMDGTWVAMQTSDETKWCPPDQLVNLH\n",
    "DFRSDEIEHLVVEELVKKREECLDALESIMTTKSVSFRRLSHLRKLVPGSFGKAYTIFNKTLMEADAHYKS\n",
    "VRTWNEIIPSKSCLKVGGRCHPHVNGVFFNGIILGPDGHVLIPEMQSSLLQQHMELLKSSVIPLMHPLAD\n",
    "PSTVFKEGDEAEDFVEVHLPDVYKQISGVDLGLPNWGKYVLMTAGAMIGLVLIFSLMTWCRRANRPESKQ\n",
    "RSFGGTGRNVSVTSQSGKVIPSWESYKSVGGQTRL'''.replace('\\n','')\n",
    "protein_C ='''MVPQALLFVPLLGFSLCFGKFPIYTIPDELGPWSPIDIHHLSCPNNLVVEDEGCTNLSEFSYMELKVGYI\n",
    "SAIKVNGFTCTGVVTEAETYTNFVGYVTTTFKRKHFRPTPDACRAAYNWKMAGDPRYEESLHGNPYPDYHW\n",
    "LRTVRTTKESLIIISPSVTDLDPYDKSLHSRVFPGGKCSGITVSSTYCSTNHDYTIWMPENPRPRTPCDI\n",
    "FTNSRGKRASKGKTCGFVDERGLYKSLKGACRLKLCGVLGLRLMDGTWVAMQTSDETKWCPPDQLVNLH\n",
    "DFRSDEEEHLVVEELVKKREECLDALESIMTTKSVSFRRLSHLRKLVPGFGKAYTIFNKTLMEADAHYKS\n",
    "VRTWNEIIPSKGCLKVGGRCHPHVNGVFFNGIILGPDGHVLIPEMQSSLLQQHMLLKSSVIPLMHPLAD\n",
    "PSTVFKEGDEAEDFVEVHLPVDVYKQISGVDLGLPNWGKYVLMTAGAMIGLVLIFSLMTWCRRANRPESKQ\n",
    "RSFGGTGGNVSVTSQSGKVIPSWESYKSGGQTFL'''.replace('\\n','')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
